Tests for Market Actor

aarshkshah1992 opened the pull request:
Closes #465  and introduces the utils to write tests for the market actor.

@anorth Let's get this in. Will be really helpful in writing tests for the other PRs.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+
+func (h *marketActorTestHarness) generateUnVerifiedDealProposal(client, provider address.Address, startEpoch, endEpoch abi.ChainEpoch) *market.DealProposal{
+	buf := make([]byte, binary.MaxVarintLen64)
+	binary.PutVarint(buf, int64(rand.Int()))
```
Please don't use randomness in tests, make them 100% deterministic. In this case I suggest taking a string or integer parameter for use as the deal data.

aarshkshah1992 replied:
This is done.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+func (h *marketActorTestHarness) generateUnVerifiedDealProposal(client, provider address.Address, startEpoch, endEpoch abi.ChainEpoch) *market.DealProposal{
+	buf := make([]byte, binary.MaxVarintLen64)
+	binary.PutVarint(buf, int64(rand.Int()))
+	hash, err := mh.Sum(buf, mh.SHA2_256, -1)
```
Use `DefaultCidBuilder` from  support/testing/cid.go

aarshkshah1992 replied:
This is done.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+	pieceSize := abi.PaddedPieceSize(2048)
+	storagePerEpoch := big.NewInt(int64(rand.Intn(1000) + 1))
+	clientCollateral := big.NewInt(int64(rand.Intn(1000) + 1))
+	providerCollateral := big.NewInt(int64(rand.Intn(1000) + 1))
```
No randoms please. I suggest setting these all zero here, and tests can mutate the result if they need a price.

aarshkshah1992 replied:
This is done.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+	require.Nil(h.t, ret)
+}
+
+func (h *marketActorTestHarness) mustGetDealProposal(rt *mock.Runtime, dealID abi.DealID) *market.DealProposal {
```
Nit: we can shorten names for these common things. I suggest `getProposal`, `getDealState` etc

aarshkshah1992 replied:
This is done.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -274,27 +271,119 @@ func TestMarketActor(t *testing.T) {
+			// create the deal to publish and the epochs around it
+			startEpoch := abi.ChainEpoch(10)
+			endEpoch := abi.ChainEpoch(20)
+			currentEpoch := abi.ChainEpoch(5)
```
`publishEpoch`

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -274,27 +271,119 @@ func TestMarketActor(t *testing.T) {
+			actor.addParticipantFunds(rt, client, deal.ClientBalanceRequirement())
+			actor.addProviderFunds(rt, provider, owner, worker, deal.ProviderBalanceRequirement())
+
+			// publish the deal so that provider collateral is locked
```
Since the deal always locks both sides, combine this with the above test. There's nothing distinct about the set-up between the two.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -274,27 +271,119 @@ func TestMarketActor(t *testing.T) {
+			// create the deal to publish and the epochs around it
+			startEpoch := abi.ChainEpoch(10)
+			endEpoch := abi.ChainEpoch(20)
+			currentEpoch := abi.ChainEpoch(5)
```
Move these epochs outside the `t.Run` if they are shared by many tests.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -274,27 +271,119 @@ func TestMarketActor(t *testing.T) {
+			st = actor.mustGetDealState(rt, dealID)
+			require.EqualValues(t, currentEpoch+1, st.SlashEpoch)
+
+			// provider cannot withdraw any funds since all it's balance is locked
```
We need more tests later than when the deal state is updated by cron, the miner can withdraw the payment for the epochs during which the deal was honoured, and that the client only recovers the amount for the rest. But this can wait for your re-work of slashing.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+
+	params := market.WithdrawBalanceParams{
+		ProviderOrClientAddress: provider,
+		Amount:                  big.Add(withDrawAmt, big.NewInt(1)),
```
This is a weird thing to have inside the harness method. We really shouldn't be doing anything tricky here. Could you move this out to the call site/s for which it's relevant?

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+	rt.Verify()
+}
+
+func (h *marketActorTestHarness) publishDealOK(rt *mock.Runtime, deal *market.DealProposal, owner, worker, provider address.Address) abi.DealID {
```
We'll very likely need this to be publish deals (plural) for subsequent tests, so just do that now.

aarshkshah1992 replied:
@anorth I'd like to defer this till we've discussed how to refactor the mock Runtime to allow multiple signature verifications.

anorth commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+		endEpoch, storagePerEpoch, providerCollateral, clientCollateral}
+}
+
+func (h *marketActorTestHarness) activeDealOK(rt *mock.Runtime, dealID abi.DealID, sectorExpiry abi.ChainEpoch, minerAddr address.Address) {
```
`activateDeals`. Similarly, make it plural now.

aarshkshah1992 replied:
This is done.

aarshkshah1992 commented on actors/builtin/market/market_test.go:
```diff
@@ -1,12 +1,16 @@
+	"bytes"
 	"context"
+	"encoding/binary"
+	acrypto "github.com/filecoin-project/specs-actors/actors/crypto"
```
This is done.

aarshkshah1992 commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
+	return dealId
+}
+
+func (h *marketActorTestHarness) generateUnVerifiedDealProposal(client, provider address.Address, startEpoch, endEpoch abi.ChainEpoch) *market.DealProposal{
```
This is done.

aarshkshah1992 commented on actors/builtin/market/market_test.go:
```diff
@@ -350,3 +439,135 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
 	)
 }
+
+func (h *marketActorTestHarness) withdrawStorageMinerBalanceOK(rt *mock.Runtime, owner, worker, provider address.Address, withDrawAmt, expectedSend abi.TokenAmount) {
```
This is done.

aarshkshah1992 commented:
@anorth I think I've addressed all your comments except https://github.com/filecoin-project/specs-actors/pull/557/#discussion_r447493475. Let's discuss that one in a bit more detail. The main problem is that the mock runtime allows only one signature verification.

aarshkshah1992 commented on actors/builtin/market/market_test.go:
```diff
@@ -470,3 +424,145 @@ func (h *marketActorTestHarness) expectProviderControlAddressesAndValidateCaller
 	)
 }
+
+func (h *marketActorTestHarness) withdrawProviderBalance(rt *mock.Runtime, withDrawAmt, expectedSend abi.TokenAmount) {
```
This has been discussed offline and @anorth is happy with this as is because the two methods set different expectations.

aarshkshah1992 commented on actors/builtin/market/market_test.go:
```diff
@@ -403,24 +360,21 @@ func TestMarketActorDeals(t *testing.T) {
+
+	owner    address.Address
+	worker   address.Address
+	provider address.Address
```
Will cover this in a follow-up PR when we add tests for multiple providers.
